### ISSUE TITLE
Safe Mode: Limit collection size 

### ISSUE THREAD

**Issue by slivne** [scylladb/scylla#10070]
Today there are no limits on collections size.

Some collections like list do not require a read before write on append.

As there is no read-before - write a workload can create lists that we will always fail to read

Unfortunately, there are internal processes that force a read before write as a background process (Materialized Views, Secondary Indexes, Repair, Compaction) - these processes will fail to read such rows. 

Add a method to limit collection size.

**Comment by nyh** [comment]
@avikivity wrote about this in #2224 (which is an abandoned issue and perhaps should be closed?), explaining why limiting collection size is not possible in the naive way: _"It's impossible to limit the size, because writes to sets do not read the set first. So you can create a huge set by adding one element at a time, and the overflow would only be detected at compaction time. We can't fail then, because all the writes have been acknowledged to the client."_.

One idea I remember we raised in the past (but can't find a link to paste here) is the following:

We already have a mechanism (https://docs.scylladb.com/troubleshooting/large-rows-large-cells-tables/) to discover oversized CQL rows during compaction - and list them in a special table (this **warning** is also a type of "safe mode", by the way). Now we can optionally (when the restriction is set to "error" mode) make it an **error to use** this row: For example, we can make it an error to write to this partition any more (more on this below) to it can't grow any more, but allow to continue reading it.

Some more detail about my idea:
1. While we want to forbid writes to an oversized row that may grow it even further, we should allow reads to it (to allow a user to salvage data saved in this row!) and should probably also allow writes which are guaranteed to shrink the row, namely deleting the entire row. Smaller tombstones (deleting individual cells or ranges) should perhaps be forbidden as well because may in the worst case cause the row to grow (if the user writes a lot of deletions to data which isn't there!).
2. The "large row table" is per-node, so each node will detect the oversized row at a different time depending on the node's compaction schedule. So if the large-row-checking is done on replicas, this may mean that a write may succeed on some of the replicas but refused on others, leading to unnecessary data inconsistencies that need to be repaired later.
3. We will need to check every write whether it tries to write to an oversized row, so we obviously can't check the on-disk large-row table every time. We'll probably need to copy of this table in memory - either as an "in-memory" table (memtable which we mark as full so we don't go to disk when something is missing in the memtable) or as a new data structure. We'll need to think what to do if this in-memory list grows too much (e.g., imagine the workload is so broken that we now have millions of oversided rows) - perhaps at that point we need to close the entire table to writes or just degrade the write performance by reading the oversized-row table from the regular (potentially not in memory) table.
4. This technique can significantly **overshoot** the desired size limit. For example if the configured limit is 1 MB and compaction waits for 4 sstables, we can have 0.9 MB in each one and compaction will result in 3.6 MB. It can actually be worse because we can have 0.9 MB on more than one size tiers (this is the same problem we always have with space amplification - see https://www.scylladb.com/2018/01/17/compaction-series-space-amplification/). But I think we can consider this reasonable: When someone will set the limit to 1MB and 1MB is fine, then 5MB is also reasonable - but we'll completely avoid the possibility of 100MB.

P.S. another related issue is #7020 where a user proposed not not making large sets an error, but rather trimming them at a fixed maximal size, leaving only the newest items.

**Comment by avikivity** [comment]
Even with all the limitations, it's better than what we have now.

**Comment by Bessonov** [comment]
Sorry for hijacking the topic. I'm not scylla user yet, but it seems like scylla could be a reasonable choice for me. I have hard time to find scylla's limits or rules of thumbs. My use case is storage of relatively small blobs (encrypted json objects). A part of data structure is:
- **id**: primary key of uuidv4 type. Could be another type, but uuid is used for the whole application until now.
- **updatedat**: used for sorting and filtering key to fetch only changed rows
- **users**: a set of user ids who can access the row
- **data**: blob of encrypted data

Could someone provide a link to the documentation or to limit benchmarks? For example I'm interested in:
- Are there any rules how many `uuids` in `users` column can be stored?
- Are there any rules how big `data` is allowed to be?
- Are there any limits on how many rows it could be?
- Are there any relevant performance drops? (Of course it's clear that the more data needs to retrieve and stored the more slower it is)

Some order of magnitude would be very helpful! Thank you in advance!

**Comment by mykaul** [comment]
@Bessonov you might find our Slack channel or discourse forums a better venue for such questions.

**Comment by Bessonov** [comment]
@mykaul thank you for you suggestion. But TBH I'm lost. I see github, I found two google groups, now there are slack and forum...

**Comment by nyh** [comment]
@Bessonov you're right that we're lacking documentation on Scylla's limits and rules of thumbs - we have an open issue about it - https://github.com/scylladb/scylladb/issues/11188. 

One thing you have to remember is that beyond the question of Scylla's implementation that may determine whether the reasonable limit is a thousand or a million items per collection (and you're right, this should be documented!), there is an inherent limitation in the CQL which makes very large collections in a single column - or a very large blob as a single value - unusable. The problem is the CQL API forces you to read the entire collection or blob in one request - there is no API (as far as I know, correct me if I'm wrong...) for paging through a collection or through an individual very long blob. So no matter how efficient Scylla is, a 1 GB collection or blob will never be usable because it will mean you need to read the entire 1 GB in one request, which rarely makes sense.

If you need very long lists or very long blobs, the more idiomatic CQL is to split them multiple items: For example you can replace single row with a long list with a partition with many rows, or replace a single huge blob or string by a partition of many smaller blobs or strings (I gave an example of this in https://stackoverflow.com/questions/74131364/key-value-database-with-random-value-access).

@Bessonov it's good that Scylla developers is active on many forums (we also try to answer stackoverflow questions :-)), but in each case you should pick the venue that makes most sense. Github issues (here) wasn't a good choice because (as you said) you had to "hijack" a topic. In my opinion Slack is almost always the wrong choice (sorry @mykaul, that's my opinion), especially because it's not searchable. The mailing lists are currently, in my opinion, the best place for such questions. You would have also gotten the attention there of more people than you did here (most Scylla developers do not look at every github issue). If you want to continue this discussion, let's do it on the Scylla mailing list please.

**Comment by ptrsmrn** [comment]
We have something similar [proposed](https://docs.google.com/document/d/11OI8SktNVZKfwo_KQB6mmHUMfXYeevGm4whSYxmgE_A/edit?usp=sharing) :
<b style="font-weight:normal;" id="docs-internal-guid-409ccdd7-7fff-c843-77c2-8ca4dbdb8514"><div dir="ltr" style="margin-left:0pt;" align="left">
Parameter | Valid range | Comments
-- | -- | --
Number of items per Collection | 20 | Soft limit to prevent creating collections with too many items
</div></b>
